### PR TITLE
Fix compilation error and some warning in Visual Studio 16.8.2.

### LINKED ISSUE #706: Failed build on Windows.
Hi all,

I've had success with installing Simbody on a few environments, but I'm getting stuck on trying to compile on my local Windows machine. Any help would be greatly appreciated!

**Basic information**:
* [Up-to-date Simbody](https://github.com/simbody/simbody/commit/dda623d3e099b38ef9e551a6b647cb5e8dfbbb25)
* CMake 3.19.0-rc3
* Visual Studio 16 2019
* RelWithDebInfo

**CMake Output**:
```
CMake Deprecation Warning at CMakeLists.txt:19 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.19042.
The C compiler identification is MSVC 19.28.29334.0
The CXX compiler identification is MSVC 19.28.29334.0
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
Detecting C compile features
Detecting C compile features - done
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
Detecting CXX compile features
Detecting CXX compile features - done
Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
Found OpenGL: opengl32   
Configuring done
Generating done
```

**Snippet of Visual Studio Output** (focusing on the error):
```
Rebuild started...
1>------ Rebuild All started: Project: ZERO_CHECK, Configuration: RelWithDebInfo x64 ------
1>Checking Build System
2>------ Rebuild All started: Project: PlatformFiles_subdir, Configuration: RelWithDebInfo x64 ------
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libblas.dll -> C:\LIB\simbody\build\RelWithDebInfo\libblas.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libgcc_s_sjlj-1.dll -> C:\LIB\simbody\build\RelWithDebInfo\libgcc_s_sjlj-1.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libgfortran-3.dll -> C:\LIB\simbody\build\RelWithDebInfo\libgfortran-3.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\liblapack.dll -> C:\LIB\simbody\build\RelWithDebInfo\liblapack.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libquadmath-0.dll -> C:\LIB\simbody\build\RelWithDebInfo\libquadmath-0.dll
2>Building Custom Rule C:/LIB/simbody/Platform/CMakeLists.txt
3>------ Rebuild All started: Project: Code - Platform Files, Configuration: RelWithDebInfo x64 ------
3>Building Custom Rule C:/LIB/simbody/CMakeLists.txt
4>------ Rebuild All started: Project: Code - SimTKcommon, Configuration: RelWithDebInfo x64 ------
4>Building Custom Rule C:/LIB/simbody/SimTKcommon/sharedTarget/CMakeLists.txt
4>About.cpp
4>NiceTypeName.cpp

...
...
...

5>Geo_Point.cpp
5>Geo_Sphere.cpp
5>Geo_Triangle.cpp
5>Geodesic.cpp
5>OBBTree.cpp
5>OrientedBoundingBox.cpp
5>gcvspl.cpp
5>About.cpp
5>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include\algorithm(7417,1): error C2678: binary '*': no operator found which takes a left-hand operand of type 'const _BidIt' (or there is no acceptable conversion)
5>        with
5>        [
5>            _BidIt=SimTK::VectorIterator<SimTK::Real,SimTK::VectorBase<SimTK::Real>>
5>        ] (compiling source file C:\LIB\simbody\SimTKmath\Geometry\src\ContactGeometry_TriangleMesh.cpp)

...
...
...



31>TestScalar.cpp
29>TestPolygonalMesh.cpp
35>TestXml.cpp
16>TestArray.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\TestArray.exe
26>TestParallelExecutor.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\TestParallelExecutor.exe
11>RandomTest.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\RandomTest.exe
8>MatVecTest.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\MatVecTest.exe
5>Done building project "SimTKmath.vcxproj" -- FAILED.

```

Any help would be greatly appreciated. I can provide more information if needed.

### PR DESCRIPTION
**Pull request by sherm1**
Visual Studio 2019 16.8.2 introduced a compilation error, fixed here. Might be a legitimate complaint.
There were also two new warnings which look spurious but are also fixed here.

Resolves #706.

cc @jasonmeverett @aymanhab

---
This change is [Reviewable](https://reviewable.io/reviews/simbody/simbody/707)

### CODE REVIEW

**Review by sherm1 — COMMENTED** [review]
+@aymanhab can you review this PR?

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707#-:-MNkN-Ab964djRWn_3Yo:bqxu0tv)* status: 0 of 3 files reviewed, all discussions resolved (waiting on @aymanhab)

**Review by aymanhab — APPROVED** [review]
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707)* status: all files reviewed, 1 unresolved discussion (waiting on @sherm1)

___
*[SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h, line 83 at r1](https://reviewable.io/reviews/simbody/simbody/707#-MNkM2FnDemRsESQ6s7n:-MNkM2Fo3sYPTsvsykHF:b3dzqti) ([raw file](https://github.com/simbody/simbody/blob/c3e65eef20edc33eab48ccd42ea85dea4b1a48df/SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h#L83)):*
> ```objc
>         return (*vectorp)[(int)index];
>     }
> ```

Thanks @sherm1 LGTM.  Any reason to remove the operator[] or just you don't expect it to be used?

**Review by sherm1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707)* status: all files reviewed, 1 unresolved discussion (waiting on @aymanhab)

___
*[SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h, line 83 at r1](https://reviewable.io/reviews/simbody/simbody/707#-MNkM2FnDemRsESQ6s7n:-MNkNCMEC_VloCnmB-sC:b-qf0j34) ([raw file](https://github.com/simbody/simbody/blob/c3e65eef20edc33eab48ccd42ea85dea4b1a48df/SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h#L83)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

Thanks @sherm1 LGTM.  Any reason to remove the operator[] or just you don't expect it to be used? 
</blockquote></details>

I think it was ill-conceived -- doesn't really make sense for an iterator since it was written to index from the beginning of the underlying vector rather than relative to the pointed-to element. Also it was apparently never used and never tested since nothing happened when I removed it!

Please let me know if OpenSim was depending on that -- hopefully not.

**Review by aymanhab — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @aymanhab and @sherm1)

___
*[SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h, line 83 at r1](https://reviewable.io/reviews/simbody/simbody/707#-MNkM2FnDemRsESQ6s7n:-MNkWtTG5w4ubsF8NaWP:bfvloyf) ([raw file](https://github.com/simbody/simbody/blob/c3e65eef20edc33eab48ccd42ea85dea4b1a48df/SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h#L83)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I think it was ill-conceived -- doesn't really make sense for an iterator since it was written to index from the beginning of the underlying vector rather than relative to the pointed-to element. Also it was apparently never used and never tested since nothing happened when I removed it!

Please let me know if OpenSim was depending on that -- hopefully not.
</blockquote></details>

Nope, we never used it, I was just curious. Thanks for the explanation.

**Review by sherm1 — COMMENTED** [review]
@chrisdembia can you help me fix continuous_integration.yml ? Apparently GitHub deprecated the `::add-path::` command and I'm not sure how to get it to work now. The deprecation instructions give a temporary workaround via an environment variable but I haven't been able to get that to work -- probably better just to use the new method but I haven't been able to figure out what that is.

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707#-:-MNkWxv91fzflvJZM48f:bjhk68r)* status: 3 of 4 files reviewed, all discussions resolved (waiting on @aymanhab)

File: .github/workflows/continuous_integration.yml
Comment (aymanhab) [inline .github/workflows/continuous_integration.yml]: @sherm1 We actually moved opensim actions to the more recent syntax, please let me know if you'd want me to update this file to use the new syntax as well.

**Review by sherm1 — COMMENTED** [review]
Here is a link to the error: https://github.com/simbody/simbody/runs/1502132557

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707#-:-MNkXbYG1bshZAZ33j4E:b-bjgy2y)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @aymanhab and @sherm1)

**Comment by aymanhab** [comment]
Thanks @sherm1 I created a branch with the fix, once done building, will merge it here.

**Review by sherm1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @aymanhab)

___
*[.github/workflows/continuous_integration.yml, line 29 at r2](https://reviewable.io/reviews/simbody/simbody/707#-MNkRX2x8H6EwgXTZ2LI-r2-29:-MNkYjJq2bYApZV3g6Cq:bpwc42g) ([raw file](https://github.com/simbody/simbody/blob/7d7f408602e9c531c109970468500cf2c95a30ca/.github/workflows/continuous_integration.yml#L29)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

@sherm1 We actually moved opensim actions to the more recent syntax, please let me know if you'd want me to update this file to use the new syntax as well.
</blockquote></details>

Oh, that would be great @aymanhab ! I'm having trouble figuring out what to do. Please feel free to push the change directly to this branch, or if you want to do a separate PR we can merge that first. I'd like this one to build cleanly before it merges.

**Comment by sherm1** [comment]
Oh, looks like I could just copy from OpenSim. I'll take a shot at that ...

**Review by sherm1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @aymanhab)

___
*[.github/workflows/continuous_integration.yml, line 29 at r2](https://reviewable.io/reviews/simbody/simbody/707#-MNkRX2x8H6EwgXTZ2LI-r2-29:-MNk_oIs9DoPooaJDOdH:b5zuz9x) ([raw file](https://github.com/simbody/simbody/blob/7d7f408602e9c531c109970468500cf2c95a30ca/.github/workflows/continuous_integration.yml#L29)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Oh, that would be great @aymanhab ! I'm having trouble figuring out what to do. Please feel free to push the change directly to this branch, or if you want to do a separate PR we can merge that first. I'd like this one to build cleanly before it merges.
</blockquote></details>

Done. Thanks, Ayman!

**Comment by aymanhab** [comment]
@sherm1 I went ahead and merged my branch to this PR since ci was past installing doxygen which was failing on windows, I hope I didn't step on your changes..

**Review by sherm1 — COMMENTED** [review]
Thanks -- perfect. 
BTW Reviewable is still waiting for a formal `:lgtm:` from you before I can merge. When you're ready ...

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707#-:-MNkaBXy45-KK7maTYC0:b-xdx7f3)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @aymanhab)

___
*[.github/workflows/continuous_integration.yml, line 29 at r2](https://reviewable.io/reviews/simbody/simbody/707#-MNkRX2x8H6EwgXTZ2LI-r2-29:-MNk_TRk2lB0mU1Q1PLS:bmjk0a3) ([raw file](https://github.com/simbody/simbody/blob/7d7f408602e9c531c109970468500cf2c95a30ca/.github/workflows/continuous_integration.yml#L29)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Done. Thanks, Ayman!
</blockquote></details>

BTW your comment here is blocking, Ayman. Please mark it satisfied (when you are).

**Review by sherm1 — COMMENTED** [review]
@chrisdembia NVM -- Ayman set me straight!

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707#-:-MNkd8CRCriABXErtzfW:baiga4q)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @aymanhab)

**Comment by chrisdembia** [comment]
👍

**Review by aymanhab — APPROVED** [review]

**Review by aymanhab — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707#-:-MNkkNI96V5J9U78MUq8:bnfp4nl)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @aymanhab)

**Review by aymanhab — APPROVED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @sherm1)

**Review by aymanhab — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/707#-:-MNktrDZ2YQ8qxayC68Y:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @sherm1)

**Comment by sherm1** [comment]
Thanks, @aymanhab !

**Pull request merged**
